src-ghc/Pact/Types/Crypto.hs

```diff
@@ -217,65 +170,73 @@ instance ConvertBS Ed25519.Signature where
   fromBS = Right . Ed25519.Sig
 #endif
 
-
-
---------- SCHEME ETH INSTANCES --------
-
-instance Scheme (SPPKScheme 'ETH) where
-  type PublicKey (SPPKScheme 'ETH) = ECDSA.PublicKey
-  type PrivateKey (SPPKScheme 'ETH) = ECDSA.PrivateKey
-  type Signature (SPPKScheme 'ETH) = ECDSA.Signature
-
-
-  _sign _ (Hash msg) pub priv = ECDSA.signETH (fromShort msg) pub priv
-  _valid _ (Hash msg) pub sig = ECDSA.validETH (fromShort msg) pub sig
-  _genKeyPair _ = ECDSA.genKeyPair
-  _getPublic _ = Just . ECDSA.getPublicKey
-  _formatPublicKey _ p = ECDSA.formatPublicKeyETH (toBS p)
-
-
-
-instance ConvertBS (ECDSA.PublicKey) where
-  toBS = ECDSA.exportPublic
-  fromBS s = maybeToEither ("Invalid ECDSA Public Key: " ++ show (toB16Text s))
-             (ECDSA.importPublic s)
-instance ConvertBS (ECDSA.PrivateKey) where
-  toBS = ECDSA.exportPrivate
-  fromBS s = maybeToEither ("Invalid ECDSA Private Key: " ++ show (toB16Text s))
-             (ECDSA.importPrivate s)
-instance ConvertBS (ECDSA.Signature) where
-  toBS = ECDSA.exportSignature
-  fromBS s = maybeToEither ("Invalid ECDSA Signature: " ++ show (toB16Text s))
-             (ECDSA.importSignature s)
-
-
-
+instance Scheme (SPPKScheme 'WebAuthn) where
+
+  _valid _ msg (PubBS pubBS) (SigBS sigBS) =
+    case runVerification of
+      Left e -> error e
+      Right () -> True
+    where
+      -- Verifying a WebAuthn signature requires that we know the payload
+      -- signed by the WebAuthn keys on the client device. This payload is
+      -- a combination of the Challenge (in our case, the PactHash of a
+      -- transaction), a JSON object of "client metadata", and data about
+      -- the authenticator hardware.
+      -- We require this data to be part of our WebAuthn signature so that
+      -- we can reconstitute the payload that was signed on the browser.
+      runVerification :: Either String ()
+      runVerification = do
+
+        -- Decode our WebAuthn signature object from a `UserSig` string.
+        WebAuthnSignature
+          { authenticatorData
+          , signature
+          , clientDataJSON } <- A.eitherDecode (BSL.fromStrict sigBS)
+        -- Decode the signer's public key.
+        publicKeyBytes <- return pubBS -- Base16.decode (T.encodeUtf8 pubBS)
+        publicKey <- first show $ Serialise.deserialiseOrFail @WA.CosePublicKey (BSL.fromStrict publicKeyBytes)
+
+        -- Recover the signature, clientData, and authData bytestrings.
+        -- Lenient decoding is not a security risk because security comes
+        -- from valitating the signature against the payload. Invalid
+        -- bytestring encodings (ones that pass lenient decoding but fail
+        -- strict decoding) will parse, but fail signature validation.
+        let
+          sig = Base64.decodeLenient (T.encodeUtf8 signature)
+          clientData = Base64URL.decodeLenient (T.encodeUtf8 clientDataJSON)
+          authData = Base64.decodeLenient (T.encodeUtf8 authenticatorData)
```

Comment:
Lenient decoding is nice, but it is *very* inefficient. Are we not allowed to expect canonical encodings for this data?

Reply:
Ooh I didn't know that, thanks! Let me double-check that lenient is required.

Reply:
I reread the spec and did some googling. Clients are indeed expected to sent valid base64 strings, so I've changed this decoding to non-lenient.